x-pack/plugins/visualization_lens/public/app.tsx

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { I18nProvider } from '@kbn/i18n/react';
+import { IScope } from 'angular';
+import React, { useCallback } from 'react';
+import { render, unmountComponentAtNode } from 'react-dom';
+import chrome from 'ui/chrome';
+
+import { PLUGIN_ID } from '../common';
+
+import { editorFrame } from '.';
+
+// Side effect of loading this is to register
+import './indexpattern_datasource';
+
+function Lens() {
+  const renderFrame = useCallback(node => {
+    if (node !== null) {
+      editorFrame.render(node);
+    }
+  }, []);
+
+  return (
+    <I18nProvider>
+      <div>
+        <h1>Lens</h1>
+
+        <div ref={renderFrame} />
+      </div>
+    </I18nProvider>
+  );
+}
+
+// TODO: Convert this to the "new platform" way of doing UI
+function App($scope: IScope, $element: JQLite) {
+  const el = $element[0];
+  $scope.$on('$destroy', () => unmountComponentAtNode(el));
+
+  return render(<Lens />, el);
+}
+
+chrome.setRootController(PLUGIN_ID, App);
```

Comment:
you'd need to test this as i'm not 100% sure it works with the legacy plugin lifecycle, but you might consider calling `chrome.setRootController` from the setup method of the plugin definition instead (`EditorFrame` class). this would more closely mirror how this will work in new platform when we have the core `application` service for registering an app.

similarly, you might be able to call `editorFrame.registerDatasource(indexPatternDatasource)` in the setup method too.

Reply:
The EditorFrame should not know anything about the app that renders it, since EditorFrame is meant to be an embeddable component. But maybe your point is that there should be a single entry point in the public folder that registers:

* The app
* The EditorFrame
* All datasources and visualizations

Reply:
Yeah I think that's what I was getting at. Of course, this is complicated a bit by the fact that this app is supposed to be temporary anyway. But the general convention used in core has been having all setup-related tasks originate from the plugin definition so that all operations related to a plugin's lifecycle live in one place.